File: desktop/src/types/ingestion.ts
export type DataType = 'orders' | 'customers' | 'products' | 'ad_spend' | 'reviews' | 'stock_levels';
export type SourceType = 'file' | 'google_sheets' | 'sample' | 'shopify_api' | 'google_sheets_api' | 'webhook' | 'postgresql';
export type IndustryTemplate = 'general' | 'fashion' | 'electronics' | 'home_garden' | 'beauty' | 'food' | 'health' | 'sports' | 'toys' | 'automotive';
export type HealthTier = 'healthy' | 'needs_review' | 'poor';

export interface ColumnMapping {
  source_column: string;
  target_column: string | null;
  auto_mapped: boolean;
  data_type_hint: string;
}

export interface DataPreview {
  headers: string[];
  rows: (string | number | null)[][];
  total_rows: number;
}

export interface DetectedSchema {
  data_type: DataType;
  confidence: number;
  suggested_mappings: ColumnMapping[];
  detected_template: IndustryTemplate;
}

export type FileImportStatus = 'pending' | 'uploading' | 'ready' | 'importing' | 'success' | 'error' | 'cancelled';

export interface UploadFileResponse {
  preview: DataPreview;
  detection: DetectedSchema;
  file_path: string;
  file_hash: string;
  mixed_types?: string[] | null;
  warnings?: string[];
}

export interface HealthCheckDetail {
  check: string;
  score: number;
  weight: number;
  issues: string[];
}

export interface GoogleSheetsRequest {
  url: string;
  sheet_name?: string;
  sheet_names?: string[];
  store_id: string;
}

export interface BatchFileItem {
  id: string;
  fileName: string;
  preview: DataPreview;
  detection: DetectedSchema;
  file_path: string;
  file_hash: string;
  status: FileImportStatus;
  error?: string | null;
  mixed_types?: string[] | null;
  warnings?: string[];
}

export interface DatabaseTableMapping {
  table: string;
  data_type: DataType;
}

export interface ImportHistoryItem {
  id: string;
  store_id: string;
  data_type: string;
  source_type: string;
  source_name: string | null;
  row_count_raw: number;
  row_count_new: number;
  row_count_updated: number;
  row_count_skipped: number;
  health_score: number;
  health_details: HealthCheckDetail[];
  status: string;
  error_message: string | null;
  imported_at: string;
  duration_ms: number | null;
}

export interface ImportLineageStep {
  id: string;
  step_order: number;
  description: string;
  rows_before: number;
  rows_after: number;
  timestamp: string;
}

export interface ColumnProfile {
  name: string;
  dtype: string;
  null_count: number;
  null_pct: number;
  unique_count: number;
  min: string | number | null;
  max: string | number | null;
  mean: number | null;
  std: number | null;
  top_values: { value: string; count: number }[];
}

export interface DataProfileResponse {
  columns: ColumnProfile[];
  row_count: number;
  date_range: { min: string; max: string } | null;
}

export function getHealthTier(score: number): HealthTier {
  if (score >= 80) return 'healthy';
  if (score >= 50) return 'needs_review';
  return 'poor';
}

export function getHealthColor(tier: HealthTier): string {
  switch (tier) {
    case 'healthy': return 'var(--novem-success)';
    case 'needs_review': return 'var(--novem-warning)';
    case 'poor': return 'var(--novem-error)';
  }
}

// ── Schema Reference ─────────────────────────────────────────────────

export interface SchemaColumnInfo {
  column: string;
  required: boolean;
  data_type: string;
  default_value: string | null;
  example: string;
  notes: string;
}

export interface SchemaReferenceResponse {
  data_type: string;
  columns: SchemaColumnInfo[];
}

// ── Validation / Dry Run ─────────────────────────────────────────────

export interface ValidationIssue {
  column: string;
  issue: string;
  severity: 'error' | 'warning' | 'info';
  affected_rows: number;
  sample_rows: number[];
}

export interface ValidateFileResponse {
  data_type: string;
  total_rows: number;
  valid_rows: number;
  droppable_rows: number;
  readiness_score: number;
  column_mapping_summary: Record<string, string>;
  unmapped_columns: string[];
  missing_required: string[];
  missing_recommended: string[];
  issues: ValidationIssue[];
  clean_actions_preview: string[];
  currency_preview: string[];
  status_preview: string[];
}

// ── Feature Readiness ────────────────────────────────────────────────

export interface FeatureReadiness {
  feature: string;
  description: string;
  minimum_required: string;
  current_value: string;
  status: 'ready' | 'warning' | 'not_ready';
  detail: string;
}

export interface DataReadinessResponse {
  features: FeatureReadiness[];
}

// ── Platform Guides ──────────────────────────────────────────────────

export interface PlatformGuide {
  platform: string;
  icon: string;
  steps: string;
}

File: engine/app/models/ingestion.py
from typing import Optional

from pydantic import BaseModel

from app.models.enums import DataType, ImportStatus, IndustryTemplate, MergeStrategy, SourceType

__all__ = ["DataType", "SourceType", "IndustryTemplate"]


# ── Column Mapping ──────────────────────────────────────────────────────


class ColumnMapping(BaseModel):
    source_column: str
    target_column: Optional[str] = None
    auto_mapped: bool
    data_type_hint: str


# ── File Upload & Preview ───────────────────────────────────────────────


class DataPreview(BaseModel):
    headers: list[str]
    rows: list[list]
    total_rows: int


class DetectedSchema(BaseModel):
    data_type: DataType
    confidence: float
    suggested_mappings: list[ColumnMapping]
    detected_template: IndustryTemplate


class UploadFileResponse(BaseModel):
    preview: DataPreview
    detection: DetectedSchema
    file_path: str
    file_hash: str
    mixed_types: Optional[list[str]] = None
    warnings: list[str] = []


# ── Quality Checks ──────────────────────────────────────────────────────


class HealthCheckDetail(BaseModel):
    check: str
    score: float
    weight: float
    issues: list[str]


# ── Import Requests & Responses ─────────────────────────────────────────


class ConfirmImportRequest(BaseModel):
    file_path: str
    store_id: str
    data_type: DataType
    mappings: list[ColumnMapping]
    has_header_row: bool = True
    selected_sheet: Optional[str] = None
    file_hash: Optional[str] = None
    merge_strategy: MergeStrategy = MergeStrategy.UPSERT


class ImportErrorDetail(BaseModel):
    row_number: Optional[int] = None
    column_name: Optional[str] = None
    error_type: str
    error_detail: str
    original_value: Optional[str] = None


class MergeResult(BaseModel):
    rows_new: int
    rows_updated: int
    rows_skipped: int
    rows_dropped: int = 0
    total_processed: int


class ConfirmImportResponse(BaseModel):
    import_id: str
    store_id: str
    data_type: str
    health_score: int
    health_details: list[HealthCheckDetail]
    merge_result: MergeResult
    lineage_steps: int
    warnings: list[str] = []
    import_errors: list[ImportErrorDetail] = []
    detection_confidence: Optional[float] = None


class GoogleSheetsRequest(BaseModel):
    url: str
    sheet_name: Optional[str] = None
    sheet_names: Optional[list[str]] = None
    store_id: str
    data_type: Optional[DataType] = None


class DatabaseConnectionTestRequest(BaseModel):
    engine: str
    host: Optional[str] = None
    port: Optional[int] = None
    database: str
    username: Optional[str] = None
    password: Optional[str] = None


# ── Import History ──────────────────────────────────────────────────────


class ImportHistoryItem(BaseModel):
    id: str
    store_id: str
    data_type: str
    source_type: str
    source_name: Optional[str] = None
    row_count_raw: int
    row_count_new: int
    row_count_updated: int
    row_count_skipped: int
    health_score: int
    health_details: list[HealthCheckDetail]
    status: str
    error_message: Optional[str] = None
    imported_at: str
    duration_ms: Optional[int] = None


class ImportLineageStep(BaseModel):
    id: str
    step_order: int
    description: str
    rows_before: int
    rows_after: int
    timestamp: str


# ── Data Profiling ──────────────────────────────────────────────────────


class ColumnProfile(BaseModel):
    name: str
    dtype: str
    null_count: int
    null_pct: float
    unique_count: int
    min: Optional[str | int | float] = None
    max: Optional[str | int | float] = None
    mean: Optional[float] = None
    std: Optional[float] = None
    top_values: list[dict]


class DataProfileResponse(BaseModel):
    columns: list[ColumnProfile]
    row_count: int
    date_range: Optional[dict] = None


# ── Store Data Summary ──────────────────────────────────────────────────


class StoreDataSummary(BaseModel):
    store_id: str
    orders_count: int = 0
    customers_count: int = 0
    products_count: int = 0
    ad_spend_count: int = 0
    reviews_count: int = 0
    stock_levels_count: int = 0
    date_range: Optional[dict] = None
    last_import_at: Optional[str] = None
    total_imports: int = 0


# ── Schema Reference ───────────────────────────────────────────────────


class SchemaColumnInfo(BaseModel):
    column: str
    required: bool
    data_type: str
    default_value: Optional[str] = None
    example: str
    notes: str


class SchemaReferenceResponse(BaseModel):
    data_type: str
    columns: list[SchemaColumnInfo]


# ── Validation / Dry Run ───────────────────────────────────────────────


class ValidationIssue(BaseModel):
    column: str
    issue: str
    severity: str
    affected_rows: int = 0
    sample_rows: list[int] = []


class ValidateFileRequest(BaseModel):
    file_path: str
    data_type: Optional[DataType] = None
    mappings: Optional[list[ColumnMapping]] = None
    has_header_row: bool = True
    selected_sheet: Optional[str] = None


class ValidateFileResponse(BaseModel):
    data_type: str
    total_rows: int
    valid_rows: int
    droppable_rows: int
    readiness_score: int
    column_mapping_summary: dict[str, str]
    unmapped_columns: list[str]
    missing_required: list[str]
    missing_recommended: list[str]
    issues: list[ValidationIssue]
    clean_actions_preview: list[str]
    currency_preview: list[str] = []
    status_preview: list[str] = []


# ── Feature Readiness ──────────────────────────────────────────────────


class FeatureReadiness(BaseModel):
    feature: str
    description: str
    minimum_required: str
    current_value: str
    status: str
    detail: str


class DataReadinessResponse(BaseModel):
    features: list[FeatureReadiness]
